### ISSUE TITLE
Random failures without --no-parallel

### ISSUE THREAD

**Issue by kosfar** [github/octocatalog-diff#119]
Hi there,
I face some kind of concurrency problem when I do not use --no-parallel option. My command looks like this:
`octocatalog-diff -n <hostname> --no-validate-references --no-parallel --fact-file /tmp/facts.yaml --from production --to testlints --puppet-master <my_puppet_master> --puppet-master-api-version 2 --puppet-master-ssl-ca ca.pem --puppet-master-ssl-client-cert cert.pem --puppet-master-ssl-client-key my.key --debug`

If I omit the --no-parallel option then I randomly get traces like the following:
```
W, [2017-05-22T22:31:06.430556 #25451]  WARN -- : Failed build_catalog for testlints
W, [2017-05-22T22:31:06.452614 #25451]  WARN -- : Catalog compile for production was aborted due to another failure
/home/kosfar/.rvm/gems/ruby-2.1.10/gems/octocatalog-diff-1.1.0/lib/octocatalog-diff/util/catalogs.rb:202:in `add_parallel_result': Catalog for testlints failed to compile due to errors: (OctocatalogDiff::Errors::CatalogError)
Failed to retrieve catalog from https://puppet-pri.noc.grnet.gr:8140/testlints/catalog/artifacts.grnet.gr: 400 undefined method `scan' for ["8", "8"]:Array at /etc/puppet/environments/testlints/modules/augeas/manifests/init.pp:6 on node artifacts.grnet.gr
	from /home/kosfar/.rvm/gems/ruby-2.1.10/gems/octocatalog-diff-1.1.0/lib/octocatalog-diff/util/catalogs.rb:112:in `block in build_catalog_parallelizer'
	from /home/kosfar/.rvm/gems/ruby-2.1.10/gems/octocatalog-diff-1.1.0/lib/octocatalog-diff/util/catalogs.rb:106:in `each'
	from /home/kosfar/.rvm/gems/ruby-2.1.10/gems/octocatalog-diff-1.1.0/lib/octocatalog-diff/util/catalogs.rb:106:in `build_catalog_parallelizer'
	from /home/kosfar/.rvm/gems/ruby-2.1.10/gems/octocatalog-diff-1.1.0/lib/octocatalog-diff/util/catalogs.rb:29:in `catalogs'
	from /home/kosfar/.rvm/gems/ruby-2.1.10/gems/octocatalog-diff-1.1.0/lib/octocatalog-diff/api/v1/catalog-diff.rb:34:in `catalog_diff'
	from /home/kosfar/.rvm/gems/ruby-2.1.10/gems/octocatalog-diff-1.1.0/lib/octocatalog-diff/api/v1.rb:19:in `catalog_diff'
	from /home/kosfar/.rvm/gems/ruby-2.1.10/gems/octocatalog-diff-1.1.0/lib/octocatalog-diff/cli.rb:112:in `cli'
	from /home/kosfar/.rvm/gems/ruby-2.1.10/gems/octocatalog-diff-1.1.0/bin/octocatalog-diff:34:in `<top (required)>'
	from /home/kosfar/.rvm/gems/ruby-2.1.10/bin/octocatalog-diff:23:in `load'
	from /home/kosfar/.rvm/gems/ruby-2.1.10/bin/octocatalog-diff:23:in `<main>'
	from /home/kosfar/.rvm/gems/ruby-2.1.10/bin/ruby_executable_hooks:15:in `eval'
	from /home/kosfar/.rvm/gems/ruby-2.1.10/bin/ruby_executable_hooks:15:in `<main>'
```
The puppet code line in which octocatalog-diff fails is this:
`if (versioncmp($::lsbmajdistrelease, '7') < 0) {
`

The only way to get this successful is to run octocatalog-diff with --no-parallel

Could you point me to some direction and shed more light on what --no-parallel does? Running parallel catalog builds has the advantage of increased speed but a more stable behavior is needed indeed. Thanks in advance

Platform and version information:
my OS: Ubuntu 14.04.5 LTS
my Ruby version: ruby 2.1.10p492 (2016-04-01 revision 54464)
my version of Puppet: 3.8.7-1puppetlabs1

**Comment by kpaulisse** [comment]
@kosfar please make sure you grab the latest version (1.2.0) - growing tired of random failures in the parallelization stuff that I was seeing in CI, I re-implemented all of that. It would be worthwhile to see if those changes also eliminate the specific problem you're seeing.

Here's a pointer to the code that implements it: https://github.com/github/octocatalog-diff/blob/master/lib/octocatalog-diff/util/parallel.rb

Basically, it's doing two `fork`s, one for the "from" catalog and one for the "to" catalog. Whereas, when you run with `--no-parallel` it doesn't fork, and instead it does them one at a time.

**Comment by kosfar** [comment]
@kpaulisse the issue persists after upgrading to 1.2.0 and it continues to be random. A workaround for me is to use "--retry-failed-catalog X". The sweet point is to use 3 retries where I start to see that some responses from catalog build get failed in the 1st attempt, others on 2nd but they never reach the 3 retries, so with 3 retries I do not get failed octocatalog-diff runs.

**Comment by kpaulisse** [comment]
I'd like to get the full debug output from one of your failures if possible. Could you post it here, or otherwise put it in a gist and e-mail the link to me at (my user name) at github.com?

**Comment by kosfar** [comment]
@kpaulisse thank you. Just emailed the gists.

**Comment by kosfar** [comment]
@kpaulisse  just pasted the logs on the gist. It's 3.8.5 on the puppetserver and 3.8.7 on my laptop.

**Comment by kpaulisse** [comment]
Since parallel code has been improved in the past months and this issue is old, I'm going to close it. If you are still having this problem and have more logs to post from the latest versions, please feel free to reopen.
